core/shared/src/main/scala/zio/prelude/InverseNonZero.scala

```diff
@@ -0,0 +1,930 @@
+package zio.prelude
+
+import zio.prelude.coherent.EqualInverseNonZero
+import zio.test.TestResult
+import zio.test.laws.{ Lawful, Laws }
+
+import scala.annotation.tailrec
+
+/**
+ * The `InverseNonZero` type class describes an associative binary operator for a
+ * type `A` that has an identity element and an inverse binary operator.
+ * Combining any value with itself with the inverse operator must return the
+ * identity element. The exception may be the `zero` element, where the result is not defined.
+ * For example, for multiplication of reals, one is an identity
+ * element and division is an inverse operation, because dividing any
+ * value by itself always returns one.
+ *
+ * Because `InverseNonZero` defines a binary rather than a unary operator it can be
+ * used to describe inverse operations for types that do not have inverse
+ * values. For example, the natural numbers do not have inverses because the
+ * set of natural numbers does not include negative numbers. But we can still
+ * define a subtraction operation that is the inverse of addition for the
+ * natural numbers, since subtracting a number from itself always returns
+ * zero.
+ */
+trait InverseNonZero[A] extends Identity[A] {
```

Comment:
I would say, instead of this, let's define a type class called `PartialInverse`, which has `(A, A) => Option[A]`. Then ordinary `Inverse` can extend that.

Reply:
Not saying it's a great idea but if we saw this pattern a lot, we could think about `InverseM`, `OrdM`, etc., with an "effectful" version of the ordinary type class. Then one common specialization would be with `F[_] = Option`, although there are certain cases where, for example, you can only tell if two things are equal by performing an effect.

Reply:
/cc @adamgfraser